parsing/unit_info.ml

```diff
@@ -0,0 +1,151 @@
+    prefix: string;
+    modname: string;
+    } -> (string, unit) t
+  | Prefix: {
+    source_file: string;
+    prefix: string;
+    modname: string;
+    } -> (string, unit) t
+  | Target:
+     {
+      source_file: string;
+      filename: string;
+      modname: string;
+    } -> (string, string) t
+  | Target_only: {
+        filename:string;
+        modname: string
+      } -> (unit, string) t
+
+type source_file = (string,unit) t
+type target_prefix = (string,unit) t
+type target = (string, string) t
+type target_only = (unit, string) t
+type 'a any_target = ('a, string) t
+
+let source_file (type a) (x: (string,a) t) = match x with
+  | Source {source_file; _ }
+  | Prefix { source_file; _ }
+  | Target { source_file ; _}  -> source_file
+
+let source_file_opt (type a b) (x: (a,b) t) = match x with
+  | Source {source_file; _ }
+  | Prefix { source_file; _ }
+  | Target { source_file ; _ } -> Some source_file
+  | Target_only _ -> None
+
+let filename (type a) (x:(a,string) t) = match x with
+  | Target { filename; _} | Target_only { filename; _ } ->  filename
+
+let modname (type a b) (x: (a,b) t) = match x with
+    | Source { modname; _ }
+    | Prefix { modname; _ }
+    | Target { modname; _ }
+    | Target_only { modname; _ }->  modname
+
+let prefix (type a b) (x: (a,b) t) = match x with
+    | Source { prefix; _ }
+    | Prefix { prefix; _ } -> prefix
+    | Target { filename; _}
+    | Target_only { filename; _ } ->
+        Misc.chop_extensions filename
+
+let basename_chop_extensions basename  =
+  match String.index basename '.' with
+  | dot_pos -> String.sub basename 0 dot_pos
+  | exception Not_found -> basename
+
+let modulize s = String.capitalize_ascii s
+
+let modname_from_source source_file =
+  source_file |> Filename.basename |> basename_chop_extensions |> modulize
+
+let from_source source_file =
+  let prefix = Misc.chop_extensions source_file in
+  let modname = modname_from_source prefix in
+  Source { modname; prefix; source_file }
+
+let start_char = function
+  | 'A' .. 'Z' -> true
+  | _ -> false
+
+let is_identchar_latin1 = function
+  | 'A'..'Z' | 'a'..'z' | '_' | '\192'..'\214' | '\216'..'\246'
+  | '\248'..'\255' | '\'' | '0'..'9' -> true
+  | _ -> false
+
+(* Check validity of module name *)
+let is_unit_name ~strict name =
+  String.length name > 0
+  && (not strict || start_char name.[0])
+  && String.for_all is_identchar_latin1 name
+
+let check_unit_name file =
+  if not (is_unit_name ~strict:true (modname file)) then
+    Location.prerr_warning (Location.in_file (source_file file))
+      (Warnings.Bad_module_name (modname file))
+
+let from_target_prefix ~source_file prefix =
+  let modname = modname_from_source prefix in
+  let p = Prefix { modname; prefix; source_file } in
+  check_unit_name p; p
+
+let from_target filename =
+  let modname = modname_from_source filename in
+  Target_only { modname; filename }
+
+let target_from_prefix ext (type a b) (p: (a,b) t): (a,string) t =
```

Comment:
The name is a bit strange as only the first two cases seem to use a prefix, and also the name `target_from_prefix` is uncomfortably close to `from_target_prefix`. Maybe `target_with_extension`?